src/sync/grandpa_warp_sync.rs

```diff
@@ -656,16 +681,33 @@ impl<TSrc> VirtualMachineParamsGet<TSrc> {
         code: Option<impl AsRef<[u8]>>,
         heap_pages: Option<impl AsRef<[u8]>>,
         exec_hint: ExecHint,
-    ) -> GrandpaWarpSync<TSrc> {
+    ) -> (
+        GrandpaWarpSync<TSrc>,
+        Option<(Error, Option<HostVmPrototype>)>,
+    ) {
         let code = match code {
             Some(code) => code,
-            None => return GrandpaWarpSync::Finished(Err(Error::MissingCode)),
+            None => {
+                return (
+                    GrandpaWarpSync::InProgress(
+                        InProgressGrandpaWarpSync::VirtualMachineParamsGet(self),
```

Comment:
You have to switch back to `WarpSyncRequest` or `WaitingForSources` here as well.
The parameters passed to `set_virtual_machine_params` are fully verified before the function is called. There's no way the user can call the function again with different values.